drivers/usb/device/usb_dc_usbip.c

```diff
@@ -0,0 +1,922 @@
+	if (ep_ctrl->cb == NULL) {
+		LOG_ERR("Control endpoint callback not set");
+		return -EIO;
+	}
+
+	if ((cmd->hdr.direction == USBIP_DIR_IN) ^
+	    (REQTYPE_GET_DIR(cmd->submit.bmRequestType) ==
+	     REQTYPE_DIR_TO_HOST)) {
+		LOG_ERR("Failed to verify bmRequestType");
+		return -EIO;
+	}
+
+	ep_ctrl->data_len = 8;
+	LOG_DBG("SETUP event ep 0x%02x %u", ep_idx, ep_ctrl->data_len);
+	memcpy(ep_ctrl->buf, &cmd->submit.bmRequestType, ep_ctrl->data_len);
+	ep_ctrl->cb(ep_idx, USB_DC_EP_SETUP);
+
+	if (cmd->hdr.direction == USBIP_DIR_OUT) {
+		/* Data OUT stage availably */
+		ep_ctrl->data_len = cmd->submit.length;
+		usbip_recv(ep_ctrl->buf, ep_ctrl->data_len);
+		LOG_DBG("DATA OUT event ep 0x%02x %u",
+			ep_idx, ep_ctrl->data_len);
+		ep_ctrl->cb(ep_idx, USB_DC_EP_DATA_OUT);
+	}
+
+	return 0;
+}
+
+static int usbip_handle_data(struct usbip_command *cmd)
+{
+	uint8_t ep_idx = cmd->hdr.ep;
+	struct usb_ep_ctrl_prv *ep_ctrl;
+	uint8_t ep;
+
+	if (cmd->hdr.direction == USBIP_DIR_OUT) {
+		if (ep_idx >= USBIP_OUT_EP_NUM) {
+			return -EINVAL;
+		}
+
+		ep_ctrl = &usbip_ctrl.out_ep_ctrl[ep_idx];
+		ep = ep_idx | USB_EP_DIR_OUT;
+		ep_ctrl->data_len = cmd->submit.length;
+		usbip_recv(ep_ctrl->buf, ep_ctrl->data_len);
+		LOG_DBG("DATA OUT event ep 0x%02x %u", ep, ep_ctrl->data_len);
+
+		ep_ctrl->cb(ep, USB_DC_EP_DATA_OUT);
+
+		/* Send ACK reply */
+		if (!usbip_send_common(ep, 0)) {
+			return -EIO;
+		}
+	} else {
+		if (ep_idx >= USBIP_IN_EP_NUM) {
+			return -EINVAL;
+		}
+
+		ep_ctrl = &usbip_ctrl.in_ep_ctrl[ep_idx];
+		ep = ep_idx | USB_EP_DIR_IN;
+		LOG_DBG("DATA IN event ep 0x%02x %u", ep, ep_ctrl->buf_len);
+
+		/* Send queued data */
+		if (!usbip_send_common(ep, ep_ctrl->buf_len)) {
+			return -EIO;
+		}
+
+		if (usbip_send(ep, ep_ctrl->buf, ep_ctrl->buf_len) !=
+		    ep_ctrl->buf_len) {
+			return -EIO;
+		}
+
+		LOG_HEXDUMP_DBG(ep_ctrl->buf, ep_ctrl->buf_len, ">");
+
+		/*
+		 * Call the callback only if data in usb_dc_ep_write()
+		 * is actually written to the intermediate buffer and sent.
+		 */
+		if (ep_ctrl->buf_len != 0) {
+			ep_ctrl->cb(ep, USB_DC_EP_DATA_IN);
+			usbip_ctrl.in_ep_ctrl[ep_idx].buf_len = 0;
+		}
+	}
+
+	return 0;
+}
+
+static int usbip_recv(uint8_t *buf, size_t len)
+{
+	return recv(usbip_ctrl.connfd, buf, len, MSG_WAITALL);
+}
+
+static int usbip_send(uint8_t ep, const uint8_t *data, size_t len)
+{
+	return send(usbip_ctrl.connfd, data, len, 0);
+}
+
+static bool usbip_send_common(uint8_t ep, uint32_t data_len)
+{
+	struct usbip_return rsp;
+	uint32_t ep_dir = USB_EP_DIR_IS_IN(ep) ? USBIP_DIR_IN : USBIP_DIR_OUT;
+	uint32_t ep_idx = USB_EP_GET_IDX(ep);
+	int rc;
+
+	rsp.hdr.command = htonl(USBIP_RET_SUBMIT);
+	rsp.hdr.seqnum = htonl(usbip_ctrl.connfd);
+	rsp.hdr.devid = htonl(0);
+	rsp.hdr.direction = htonl(ep_dir);
+	rsp.hdr.ep = htonl(ep_idx);
+
+	memset(&rsp.submit, 0, sizeof(rsp.submit));
+	rsp.submit.actual_length = htonl(data_len);
+
+	rc = send(usbip_ctrl.connfd, &rsp, sizeof(rsp), 0);
```

Comment:
Same here, would be better to loop over `send()` until the entire payload is sent, and break only on an error condition.